pydatastructs/trees/binary_trees.py

```diff
@@ -632,6 +636,141 @@ def _left_rotate(self, j, k):
+            parent_idx = self.tree[node_idx].parent
+            if parent_idx is not None:
+                parent = self.tree[parent_idx]
+        if node.parent is None:
+            self.tree[node_idx].is_root = True
+
+    def _trickle_down(self, node_idx):
+        node = self.tree[node_idx]
+        parent_idx = node.parent
+        while node.left is not None or node.right is not None:
+            if node.left is None:
+                self._left_rotate(node_idx, self.tree[node_idx].right)
+            elif node.right is None:
+                self._right_rotate(node_idx, self.tree[node_idx].left)
+            elif self.tree[node.left].priority < self.tree[node.right].priority:
+                self._right_rotate(node_idx, self.tree[node_idx].left)
+            else:
+                self._left_rotate(node_idx, self.tree[node_idx].right)
+            node = self.tree[node_idx]
+            parent_idx = node.parent
+
+    def insert(self, key, data, priority):
+        super(CartesianTree, self).insert(key, data)
+        node_idx = super(CartesianTree, self).search(key)
+        node = self.tree[node_idx]
+        new_node = CartesianTreeNode(key, priority, data)
+        new_node.parent = node.parent
+        new_node.left = node.left
+        new_node.right = node.right
+        self.tree[node_idx] = new_node
+        if node.is_root:
+            self.tree[node_idx].is_root = True
+        else:
+            self._bubble_up(node_idx)
+
+    def delete(self, key, **kwargs):
+        balancing_info = kwargs.get('balancing_info', False)
+        node_idx = super(CartesianTree, self).search(key)
+        if node_idx is not None:
+            self._trickle_down(node_idx)
+            return super(CartesianTree, self).delete(key, balancing_info = balancing_info)
+
+    def __str__(self):
+        to_be_printed = ['' for i in range(self.tree._last_pos_filled + 1)]
+        for i in range(self.tree._last_pos_filled + 1):
+            if self.tree[i] is not None:
+                node = self.tree[i]
+                to_be_printed[i] = (node.left, node.key, node.priority, node.data, node.right)
+        return str(to_be_printed)
+
+class Treap(CartesianTree):
+    """
+    Represents treaps.
+
+    Examples
+    ========
+
+    >>> from pydatastructs.trees import Treap as T
+    >>> t = T()
+    >>> t.insert(1, 1)
+    >>> t.insert(2, 2)
+    >>> t.search(1)
+    0
+    >>> t.search(-1) is None
+    True
+    >>> t.delete(1) is True
+    True
+    >>> t.search(1) is None
+    True
+    >>> t.delete(2) is True
+    True
+    >>> t.search(2) is None
+    True
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Treap
+
+    """
+    priorities = set()
+    def insert(self, key, data):
+        priority = random.random()
+        while priority in self.priorities:
+            priority = random.random()
+        self.priorities.add(priority)
+        super(Treap, self).insert(key, data, priority)
```

Comment:
```suggestion
    def insert(self, key, data):
        priorities = set()
        priority = random.random()
        while priority in priorities:
            priority = random.random()
        priorities.add(priority)
        super(Treap, self).insert(key, data, priority)
```
It looks like unique priorities are necessary. Is it so?

Reply:
> The structure of the tree is determined by the requirement that it be heap-ordered: that is, the priority number for any non-leaf node must be greater than or equal to the priority of its children.

It looks equal priorities are also fine.

Reply:
It is not necessary to have unique priorities, however it is better to have unique priorities, as it can ensure better that the height of the tree is proportional to logarithm of size/number of nodes. And since, `set` allows a lookup in constant time, therefore I restricted the priorities to be unique. I can work to allow equal priorities as well, if you suggest.

Reply:
> And since, `set` allows a lookup in constant time, therefore I restricted the priorities to be unique. I can work to allow equal priorities as well, if you suggest.

`priorities` is eating extra memory. And there are very low chances that, `random.random` will give same output on two different calls. So, why to create extra burden for such a rarely occurring case?

Reply:
That makes sense. Okay then I will make the changes accordingly